### PR TITLE
Tempus bug setting tableau and initializing 4965

### PR DESCRIPTION
**Pull request by ccober6**
@trilinos/tempus 

## Description
Resetting an RK tableau does not seem to work correctly.  Need to cleanup the construction and initialization of the tableaus.

### CODE REVIEW

File: packages/tempus/src/Tempus_RKButcherTableau.hpp [hunk 1]
```diff
@@ -285,9 +188,105 @@ class RKButcherTableau :
...
+      order_ = order;
+      orderMin_ = orderMin;
+      orderMax_ = orderMax;
+      this->set_isImplicit();
+      this->set_isDIRK();
+
+      // Consistency check on b
+      Scalar sumb = Scalar(0.0);
+      for (size_t i = 0; i < this->numStages(); i++) sumb += b_(i);
+      TEUCHOS_TEST_FOR_EXCEPTION( std::abs(Scalar(1.0)-sumb) > 1.0e-08,
```
Comment (mhoemmen) [inline packages/tempus/src/Tempus_RKButcherTableau.hpp]: Would you consider picking this tolerance as a function of Scalar, say by using `Teuchos::ScalarTraits`?

File: packages/tempus/src/Tempus_RKButcherTableau.hpp [hunk 2]
```diff
@@ -1678,89 +1793,28 @@ class GeneralDIRK_RKBT :
...
-
-    // Fill c:
-    c(0) = ST::one();
-
-    int order = 1;
-
-    this->initialize(A,b,c,order,this->getDescription());
-  }
+    const Scalar one   = ST::one();
+    const Scalar gamma = as<Scalar>((2*one-ST::squareroot(2*one))/(2*one));
```
Comment (mhoemmen) [inline packages/tempus/src/Tempus_RKButcherTableau.hpp]: Be careful with `operator* (int, Scalar)`.  It does not exist for complex Scalar types, for example.  See the following example: https://godbolt.org/z/kea6Im

File: packages/tempus/src/Tempus_RKButcherTableau.hpp (hunk 1 above)
Comment (ccober6) [reply packages/tempus/src/Tempus_RKButcherTableau.hpp]: Hi Mark,

Not sure if I understand you.  I probably should have used Teuchos::ScalarTraits for Scalar(1.0).  Are you suggesting using Teuchos::ScalarTraits for the 1.0e-08?

Comment (mhoemmen) [reply packages/tempus/src/Tempus_RKButcherTableau.hpp]: @ccober6 You could use `Teuchos::ScalarTraits<Scalar>::epsilon()` (or perhaps the square root of that).  `Scalar(1.0)` is fine.  (`Scalar(1)` might not be, for complex `Scalar`.)

Comment (mhoemmen) [reply packages/tempus/src/Tempus_RKButcherTableau.hpp]: It doesn't really matter if you don't test for `float`, though.

**Comment by ccober6** [comment]
Hi Mike,

` *  We have several classes that currently derive from StepperExplicitRK - they are at the same level as StepperERK_General. For example, one allows me to compute an arbitrary set of embedded error estimates, given any tableau and bstar vectors. These use StepperExplicitRK to extract information from the tableau and stepper parameter lists. Instead of a parameter list, should the constructors of these classes be modified to directly receive information like useEmbedded and the tableau details (matching what StepperERK_General is currently doing)?`

So you can do it basically two ways now.  You can still use the ParameterList, but the construction is through the StepperFactory (e.g., StepperFactory::createStepperERK_General().  Or as you mentioned above you can use one of the Stepper constructors, and use the "set*()" functions like StepperERK_General is currently doing.

` *  Were Tempus::Explicit4Stage4thOrder_RKBT and Tempus::ForwardEuler_RKBT eliminated? Is there any similarly convenient way to access their tableaus for use in a general ERK stepper? I suppose I could just build the Steppers for those types directly instead of using their tableaus in a general form. 
`

So the tableaus have been moved into separate Steppers.  So Tempus::Explicit4Stage4thOrder_RKBT and Tempus::ForwardEuler_RKBT  are now StepperERK_4Stage4thOrder and StepperERK_ForwardEuler in Tempus_StepperRKButcherTableau.hpp.  If you want the tableaus of these steppers, you can get them like ...

```
    auto stepperERK = Teuchos::rcp(new StepperERK_Trapezoidal<Scalar>());
    this->setExplicitTableau(stepperERK->getTableau());

```
in Tempus_StepperIMEX_RK_impl.hpp:142.  There other examples in there on how to build tableaus directly in Tempus_StepperIMEX_RK_impl.hpp:154.

Let me know if this does not make sense.

**Comment by michael-a-hansen** [comment]
Thanks Curt! That all makes sense. I'll be on vacation next week so I will try to make some progress on this shortly.

**Comment by michael-a-hansen** [comment]
Hello again,

I have SPARC's explicit solvers compiling against your branch. I haven't been able to run with them because the _implicit_ tempus solver in SPARC is calling `tempus_stepper_->getNonconstParameterList()` which apparently does not exist anymore. It is only using it to populate a PList for NOX, as in:

```
tempus_stepper_pl = tempus_stepper_->getNonconstParameterList();
nox_solver_pl = tempus_stepper_pl->sublist("Default Solver", true);
```

How should I get a NOX solver PList now that we can't retrieve a PList from the stepper?

**Comment by ccober6** [comment]
@ikalash Great!  It was in my plans today to do a merge, so this should work out!

**Comment by ikalash** [comment]
Sounds like a plan!

**Comment by ccober6** [comment]
@michael-a-hansen    So you can grab a default solver ParameterList via Tempus::defaultSolverParameters() in Tempus_Stepper_decl.hpp.  It is a global function, so you do not even need a stepper to call it.  I believe all the steppers use this as their default solver settings, so should be the same as what you currently have.

**Comment by sconde** [comment]
@ccober6 There's a conflict with Trilinos/develop

**Comment by ccober6** [comment]
Is it "Adding option to tempus DIRK stepper for resetting initial guess. #5694"?  Something about "Reset Initial Guess"?  I fixed that but I may not have pushed it trilinos/Trilinos.  I will try to get to it tomorrow.

**Comment by ccober6** [comment]
@sconde,  @jmgate helped cleanup this branch, but it removed the approval.  The changes are exactly the same as before so can you please re-review and approve so this PR can be merged?  
Also, you may need to re-checkout this branch, as it changed its history.

Thanks, Curt

**Comment by sconde** [comment]
@ccober6 Ok. I've pulled and building now

**Comment by jmgate** [comment]
@trilinos/framework, what does this error mean?
```
Set CWD = /home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/TFW_testing_single_configure_prototype
Traceback (most recent call last):
File "/home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 524, in 
returnValue = run()
File "/home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/cmake/std/PullRequestLinuxDriverTest.py", line 518, in run
'package_subproject_list.cmake'])
File "/home/projects/ppc64le/python/2.7.12/lib/python2.7/subprocess.py", line 541, in check_call
raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ctest', '-S', 'simple_testing.cmake', '-Dbuild_name=PR-5543-test-Trilinos_pullrequest_cuda_9.2-2096', '-Dskip_by_parts_submit=OFF', '-Dskip_update_step=ON', '-Ddashboard_model=Experimental', '-Ddashboard_track=Pull Request', '-DPARALLEL_LEVEL=64', '-Dbuild_dir=/home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/pull_request_test', '-Dconfigure_script=/home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/cmake/std/PullRequestLinuxCuda9.2TestingSettings.cmake', '-Dpackage_enables=../packageEnables.cmake', '-Dsubprojects_file=../TFW_single_configure_support_scripts/package_subproject_list.cmake']' returned non-zero exit status 255
Build step 'Execute shell' marked build as failure
Archiving artifacts
Finished: FAILURE
```

**Comment by prwolfe** [comment]
It means test failed in some way. There should be more details in the crash and Jenkins logs.

**Comment by sconde** [comment]
> Is @sconde not a Trilinos developer?

Have I been blacklisted?

**Comment by prwolfe** [comment]
I don't think @sconde has been blacklisted. The approval was on commit 86f23a5fee29fb4afbda7c0d3780049b45d92c03 and the last push at that point was bff56d98037544cd720a020ec6133ad4fa74a66c.

**Pull request merged**